### PR TITLE
Trustedmode backuprestore

### PR DESCRIPTION
**Pull request by vasanthrajams**
# Swagger file specification for supporting Trusted Mode Backup/Restore/Selective Restore - Pull Request

## API Info: The Basics
This PR contains changes required for supporting the trusted mode method of backup/restore/selective restore operations for managed hsm service. In addition to supporting the existing Azures Storage service SAS token based backup/restore and selective restore operations, with this change, managed HSM will also start supporting the backup/restore operations with just passing the storage uri.

The storage authentication type is decided based on the boolean parameter passed by the client. if useManagedIdentityStorageAuthentication is set, then the sas token is ignored. 

More detailed design specification can be found [here](https://microsoft.sharepoint.com/:w:/t/HSMvNext/EaPz0nDgef1IpMWIDkAP7FQB346PLVUK-mu6i-5uG1Rasw?e=iUZ3hS&ovuser=72f988bf-86f1-41af-91ab-2d7cd011db47%2cvachitta%40microsoft.com&OR=Teams-HL&CT=1682118119845&clickparams=eyJBcHBOYW1lIjoiVGVhbXMtRGVza3RvcCIsIkFwcFZlcnNpb24iOiIyNy8yMzA0MDIwMjcwMyIsIkhhc0ZlZGVyYXRlZFVzZXIiOmZhbHNlfQ%3d%3d&CT=1682980770546&OR=Outlook-Body&CID=17CF87CB-FCCF-4777-BF8F-8E4F8E073C4D&wdLOR=c07BC5460-3010-4480-8C1E-BD224CE52669)

Is this review for:

- [X] a private preview
- [ ] a public preview
- [ ] GA release

### CODE REVIEW

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 1]
```diff
@@ -82,6 +82,61 @@
         }
       }
     },
+    "/managedidentitymodebackup": {
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:85]: POST operations should take the form of a verb. Maybe that's what "backup" means here (more like "back up" sans spaces), but this long endpoint name isn't very descriptive nor does it follow many of our recommended patterns. For example, shouldn't this build on the existing /backup endpoint? What exactly is the "Manage Identity Mode" adding here? Isn't that just an authentication schema? Can this just be a change to the input model for /backup?

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 2]
```diff
@@ -82,6 +82,61 @@
...
+        "x-ms-long-running-operation-options": {
+          "final-state-via": "azure-async-operation"
+        },
+        "description": "Creates a full backup using a user-provided Azure blob storage container uri.",
+        "parameters": [
+          {
+            "in": "body",
+            "name": "azureStorageBlobContainerUri",
+            "schema": {
+              "type": "string"
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:101]: The body should always be a model object. You can't expand this, so in the future if you need to add properties it's a breaking change. See https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#http-post-action-returns-200

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 3]
```diff
@@ -82,6 +82,61 @@
...
+        ],
+        "responses": {
+          "202": {
+            "description": "Started full backup",
+            "headers": {
+              "Retry-After": {
+                "description": "The recommended number of seconds to wait before calling the URI specified in Azure-AsyncOperation.",
+                "type": "integer"
+              },
+              "Azure-AsyncOperation": {
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:117]: Data plane APIs should use Operation-Location: https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#obtaining-status-and-results-of-long-running-operations

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 4]
```diff
@@ -82,6 +82,61 @@
...
+    "/managedidentitymodebackup": {
+      "post": {
+        "tags": [
+          "ManagedIdentityModeFullBackup"
+        ],
+        "operationId": "ManagedIdentityModeFullBackup",
+        "x-ms-long-running-operation": true,
+        "x-ms-long-running-operation-options": {
+          "final-state-via": "azure-async-operation"
+        },
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:94]: You need to use data-plane LROs. This should be a status monitor per https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#obtaining-status-and-results-of-long-running-operations

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 5]
```diff
@@ -177,6 +232,61 @@
...
+    "/managedidentitymoderestore": {
+      "put": {
+        "tags": [
+          "ManagedIdentityModeFullRestore"
+        ],
+        "operationId": "ManagedIdentityModeFullRestore",
+        "x-ms-long-running-operation": true,
+        "x-ms-long-running-operation-options": {
+          "final-state-via": "azure-async-operation"
+        },
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:244]: Same comments here for LROs. Please following the [guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#obtaining-status-and-results-of-long-running-operations) for data-plane APIs.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 6]
```diff
@@ -278,6 +388,68 @@
...
+    "/keys/{keyName}/managedidentitymoderestore": {
+      "put": {
+        "tags": [
+          "Keys"
+        ],
+        "operationId": "ManagedIdentityModeSelectiveKeyRestoreOperation",
+        "x-ms-long-running-operation": true,
+        "x-ms-long-running-operation-options": {
+          "final-state-via": "azure-async-operation"
+        },
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:401]: Same everywhere: use data plane LRO guidelines.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 7]
```diff
@@ -278,6 +388,68 @@
           }
         }
       }
+    },
+    "/keys/{keyName}/managedidentitymoderestore": {
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:392]: Do you have suitable validation in place to make sure no key version could ever be "managedidentitymoderestore"?

**Review by heaths — CHANGES_REQUESTED** [review]

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json (hunk 2 above)
Comment (docschmidt) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:101]: In addition to that, do you want to include an optional folder path parameter?

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 8]
```diff
@@ -39,6 +39,14 @@
...
         "description": "Creates a full backup using a user-provided SAS token to an Azure blob storage container.",
         "parameters": [
+          {
+            "name": "storageAuthType",
+            "in": "query",
+            "description": "Type of authentication to use",
+            "required": false,
+            "type": "string",
+            "enum": ["sasToken", "managedIdentity"]
+          },
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:49]: Given this is a POST already, why not just add this as a request model property?

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 9]
```diff
@@ -189,6 +142,14 @@
...
         "description": "Restores all key materials using the SAS token pointing to a previously stored Azure Blob storage backup folder",
         "parameters": [
+          {
+            "name": "storageAuthType",
+            "in": "query",
+            "description": "Type of authentication to use",
+            "required": false,
+            "type": "string",
+            "enum": [ "sasToken", "managedIdentity" ]
+          },
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:152]: Why is this needed when you already know the job details from the given job ID? Normally how these endpoints are exposed won't lend themselves to be passed another parameter by our pollers.

**Review by heaths — CHANGES_REQUESTED** [review]
I strongly suggest moving the auth type parameter to the body, but the parameter to the status endpoint may not work well with code generation nor should it be necessary since it's just a status API and you should know what the job is using, I would think.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json (hunk 8 above)
Comment (vasanthrajams) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:49]: We wanted to minimize the number of changes to the already existing client code and not break them with the new api version. As a result, after discussing, we thought instead of modifying the body parameters, we thought exposing via optional query parameters helps to allow the newer client to pass the managedIdentity auth mode and existing clients just can continue to use the same request without passing any query parameter even when they upgrade their client.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json (hunk 9 above)
Comment (vasanthrajams) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:152]: This is the restore request when called 202 and as a response returns the polling uri. We only need this query parameter for backup (/backup), restore(/restore) and selective key restore (/keys/{keyName}/restore). 

The polling uri endpoints as (/backup/{jobId}/pending), (/restore/{jobId}/pending). We have not added the query parameters to these as they are not needed as you mentioned heath.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 10]
```diff
@@ -134,6 +142,14 @@
         },
         "description": "Restores all key materials using the SAS token pointing to a previously stored Azure Blob storage backup folder",
         "parameters": [
+          {
+            "name": "storageAuthType",
+            "in": "query",
+            "description": "Type of authentication to use",
```
Comment (docschmidt) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:148]: `description` is shown as part of the document for the REST API, isn't it? If so, we need to be more specific what we mean here, even though the parameter name might give a hint. 

We could say, for example: Specifies the authentication type that is used to authenticate Managed HSM with Azure Storage. The default option, `sasToken`, uses the provided SAS token for authentication/authorization. `managedIdentity` uses the user-assigned managed identity that was assigned to the managed HSM resource.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 11]
```diff
@@ -229,6 +245,14 @@
         },
         "description": "Restores all key versions of a given key using user supplied SAS token pointing to a previously stored Azure Blob storage backup folder",
         "parameters": [
+          {
+            "name": "storageAuthType",
+            "in": "query",
+            "description": "Type of authentication to use",
```
Comment (docschmidt) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:251]: Same comment as above. Be more specific in the description.

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 12]
```diff
@@ -39,6 +39,14 @@
         },
         "description": "Creates a full backup using a user-provided SAS token to an Azure blob storage container.",
         "parameters": [
+          {
+            "name": "storageAuthType",
+            "in": "query",
+            "description": "Type of authentication to use",
```
Comment (docschmidt) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:45]: See my comment below regarding `description`.

**Review by docschmidt — CHANGES_REQUESTED** [review]

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json (hunk 8 above)
Comment (heaths) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json]: Adding properties in a new API version is not a breaking change. Either way, it's a change for clients. The property - like the parameter - should be optional anyway, so how would it break existing clients?

@johanste or @JeffreyRichter would we want a POST endpoint mixing body and query parameters? We've seen it before as more of a filter, I know, but this is used as more of a directive and seems it should be in the body parameter - whether to use SAS or MSI auth along with the current storage container URI.

**Review by docschmidt — APPROVED** [review]

**Review by docschmidt — APPROVED** [review]

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 13]
```diff
@@ -376,6 +352,11 @@
         "token": {
           "type": "string",
           "description": "The SAS token pointing to an Azure Blob storage container"
+        },
+        "useManagedIdentityStorageAuthentication": {
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:356]: It's descriptive, but many of the REST API and SDK architects aren't big on long names. Perhaps something like `useManagedIdentity`? "Storage" and "Authentication" both seem extraneous given the context.

**Review by heaths — APPROVED** [review]

File: specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json [hunk 14]
```diff
@@ -352,11 +352,15 @@
...
+        },
+        "useManagedIdentity": {
+          "type": "boolean",
+          "default": false,
+          "description": "Indicates which authentication method should be used. If set to true, Managed HSM will use the configured user-assigned managed identity to authenticate with Azure Storage. Otherwise, a SAS token has to be specified."
         }
       },
       "required": [
-        "storageResourceUri",
-        "token"
```
Comment (heaths) [inline specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:359]: This is a breaking change even when changing the api-version. See https://aka.ms/AzBreakingChangesPolicy (internal only). What happens when a client sends `null` or an empty string? That's likely what is going to happen. Can you handle that on the backend to treat it the same as absent?

@mikekistler @JeffreyRichter @johanste does this need to go through the breaking change approval process, or will that depend on their answer? We can, if needed, patch this in code generated clients, but most languages are hand-coding re-export of models so even that may not be necessary.

Comment (heaths) [reply specification/keyvault/data-plane/Microsoft.KeyVault/preview/7.5-preview.1/backuprestore.json:359]: From the service team,

> When the useManagedIdentity is set, the token is optional and will be ignored at the service side.
